File: ios/Echo/Services/AI/GrammarCorrectionService.swift
import Foundation

/// Protocol defining Grammar Correction Service contract.
public protocol GrammarCorrectionServiceProtocol {
    var isEnabled: Bool { get set }
    func correctGrammar(rawTranscript: String, config: LLMConfig) async throws -> String
}

/// Service handling optional AI-powered grammar correction for transcriptions.
public class GrammarCorrectionService: GrammarCorrectionServiceProtocol {
    public static let systemPrompt = "Correct grammar, punctuation, capitalization, spelling and formatting without changing the meaning, tone or wording unnecessarily. Do not summarize. Return only the corrected text."
    
    private let llmProviderService: LLMProviderServiceProtocol
    public var isEnabled: Bool
    
    public init(
        llmProviderService: LLMProviderServiceProtocol = LLMProviderService(),
        isEnabled: Bool = false
    ) {
        self.llmProviderService = llmProviderService
        self.isEnabled = isEnabled
    }
    
    public func correctGrammar(rawTranscript: String, config: LLMConfig) async throws -> String {
        // If grammar correction is disabled or text is empty, return raw transcript untouched
        guard isEnabled || config.isGrammarCorrectionEnabled else {
            return rawTranscript
        }
        
        let trimmed = rawTranscript.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmed.isEmpty else {
            return rawTranscript
        }
        
        return try await llmProviderService.generateCompletion(
            prompt: trimmed,
            systemPrompt: GrammarCorrectionService.systemPrompt,
            config: config
        )
    }
}

File: android/app/src/main/java/com/echo/dictation/ai/GrammarCorrectionService.kt
package com.echo.dictation.ai

import android.util.Log
import javax.inject.Inject
import javax.inject.Singleton

/**
 * Applies optional AI grammar correction to a raw transcript.
 *
 * Grammar correction is optional and controlled by [enabled].
 * When disabled the service is a no-op and returns the raw text immediately —
 * the transcription pipeline continues unmodified.
 *
 * The UI must communicate with this service only through [AIService] —
 * nothing should call [GrammarCorrectionService] directly.
 */
@Singleton
class GrammarCorrectionService @Inject constructor(
    private val llm: LlmProvider,
) {

    /**
     * Whether grammar correction is currently active.
     * Toggled from [com.echo.dictation.data.local.AppPreferences].
     */
    var enabled: Boolean = false

    /**
     * Correct [rawTranscript] if [enabled] is true.
     *
     * @return [LlmResult.Success] with the corrected text when enabled and the call succeeds,
     *         [LlmResult.Success] with the original [rawTranscript] when disabled,
     *         [LlmResult.Failure] only when enabled and the LLM call fails.
     */
    suspend fun correct(rawTranscript: String): LlmResult {
        if (!enabled) {
            Log.d(TAG, "Grammar correction disabled — returning raw transcript")
            return LlmResult.Success(rawTranscript)
        }

        if (rawTranscript.isBlank()) {
            Log.d(TAG, "Empty transcript — skipping grammar correction")
            return LlmResult.Success(rawTranscript)
        }

        Log.d(TAG, "Grammar correction enabled — calling LLM (${rawTranscript.length} chars)")
        return llm.complete(
            systemPrompt = SYSTEM_PROMPT,
            userPrompt   = rawTranscript,
        )
    }

    companion object {
        private const val TAG = "GrammarCorrectionService"

        /** Exact prompt specified in the product requirements. */
        const val SYSTEM_PROMPT =
            "Correct grammar, punctuation, capitalization, spelling and formatting " +
            "without changing the meaning, wording or tone. " +
            "Do not translate the text to another language; keep the original spoken language. " +
            "Do not summarize. " +
            "Do not remove information. " +
            "Return only the corrected transcript."
    }
}
